### PR TITLE
Association

### PR DESCRIPTION
**Pull request by Thermiee**
In this pull request, I:

- Created a class `Classroom` with the following:
  - `@label` instance variable, initialized in the constructor
  - Setter and getter for `@label`
-  Created the `has-many/belongs-to` relationship between `Classroom` and `Student`. The following was implemented:
   - Created the `has-many` side (a classroom has many students)
   - Created the `belongs-to` side (a student belongs to a classroom)
- Created a class `Book` with the following:
   - `@title` and `@author` instance variables, initialized in the constructor
   - Setters and getters for instance variables
- Created a class `Rental` with the following
   - `@date` instance variable, initialized in the constructor
   - Setter and getter for `@date`
- Created the `many-to-many` (also `has-many-through`) relationship between `Person` and `Book` using the intermediate class `Rental`. The following was implemented:
  - Created the `has-many` side of `Book` and `Rental` (a book has many rentals)
  - Created the ` belongs-to` side of `Rental` and `Book` (a rental belongs to a book)
  - Created the `has-many` side of `Person` and `Rental` (a person has many rentals)
  - Created the `belongs-to` side of `Rental` and `Person` (a rental belongs to a person)
  - Modified the constructor of `Rental` so `Book` and `Person` are set in it.

### CODE REVIEW

File: person.rb [hunk 1]
```diff
@@ -1,17 +1,18 @@
...
+  attr_reader :id
+
   def initialize(age, name = 'Unknown', parent_permission: true)
     @id = Random.rand(1..1000)
     @name = name
     @age = age
     @parent_permission = parent_permission
+    @rental = []
     super()
   end
```
Comment (t-yanick) [inline person.rb:15]: - [x] To implement the `has-many` side of the relationship between `Person` and `Rental`, a method for adding rentals should be present in the `Person` class like so;
```ruby
  def add_rental(book, date)
    Rental.new(date, book, self)
  end
```
Please kindly fix that.

File: classroom.rb [hunk 2]
```diff
@@ -0,0 +1,14 @@
...
+
+  def initialize(label)
+    @label = label
+    @students = []
+  end
+
+  def add_student(student)
+    @students.push(student)
+    student.classroom = self
+  end
```
Comment (t-yanick) [inline classroom.rb:13]: - [x] To implement the `belongs-to` association between `Classroom` and `Student`, `a method that serves as the 'belongs-to' side (a student belongs to a classroom)` should be present in the `Student` class. Something like so:
```ruby
  def classroom=(classroom)
      @classroom = classroom
      classroom.students.push(self) unless classroom.students.include?(self)
  end
```
Please kindly fix that.

**Review by t-yanick — CHANGES_REQUESTED** [review]
Hi @Thermiee, 

Good job so far!
There are some issues that you still need to work on to go to the next project but you are almost there!

## Required Changes ♻️
_Check the comments under the review._

## Optional suggestions
_Every comment with the [OPTIONAL] prefix is not crucial enough to stop the approval of this PR. However, I strongly recommend you to take them into account as they can make your code better._

Cheers and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Please, remember to tag me in your question so I can receive the notification.

**_Please, do not open a new Pull Request for re-reviews. You should use the same Pull Request submitted for the first review, either valid or invalid unless it is requested otherwise._**

------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561) you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

File: book.rb [hunk 3]
```diff
@@ -0,0 +1,17 @@
...
+
+  def initialize(title, author)
+    @title = title
+    @author = author
+    @rentals = []
+  end
+
+  def add_rental(date, person, book)
+    @rentals << Rental.new(date, person, book)
+  end
```
Comment (Deepakdanger) [inline book.rb]: - [ ] [OPTIONAL] While creating rental in `class Book`. we don't require to pass  `book` here. kindly fix it .
- for ex:- 
```
 def add_rental(person, date)
    # *the book itself should be sent as a parameter to create the new rental achieving the 'has-many' association. 
    Rental.new(date, self, person)
  end
```

File: rental.rb [hunk 4]
```diff
@@ -0,0 +1,14 @@
+class Rental
+  attr_accessor :date
+  attr_reader :person, :book
```
Comment (Deepakdanger) [inline rental.rb]: - [ ] [ OPTIONAL] It is required to have `:person, :book` inside `attr_accessor`. 
- **Requirement** `attr_accessor :date, :book, :person`

File: book.rb [hunk 5]
```diff
@@ -0,0 +1,17 @@
...
+
+  def classroom=(classroom)
+    @classroom = classroom classroom.students.push(self) unless classroom.students.include?(self)
+  end
```
Comment (Deepakdanger) [inline book.rb]: - [ ] [**REQUIRED**] As requested by the previous reviewer, To implement the `belongs-to` association between `Classroom` and `Student`, a method that serves as the 'belongs-to' side (a student belongs to a classroom) should be present in the Student class.
- Kindly create a new file and write this code there, For ex:- 
- * When setting the classroom for a student it also adds it to the classrooms' students:

```ruby
class Student < Person
  attr_reader :classroom
  .
  .
  # *Check for a method that serves as the 'belongs-to' side (a student belongs to a classroom). Link to the article with examples: https://github.com/microverseinc/curriculum-ruby/blob/main/oop/articles/oop_relationships_by_examples.md#both-ways-relationships
  def classroom=(classroom)
      @classroom = classroom
      classroom.students.push(self) unless classroom.students.include?(self)
  end
end
```

**Review by Deepakdanger — CHANGES_REQUESTED** [review]
Hi @Thermiee, 

Great work on making the changes requested by a previous reviewer 👏 
You've done well implementing some of the requested changes, but there are still some which aren't addressed yet.

 ## Required Changes ♻️
- [ ] _Check the comments under the review._

## Optional suggestions

_Every comment with the [OPTIONAL] prefix won't stop the approval of this PR. However, I strongly recommend you to take them into account as they can make your code better. Some of them were simply missed by the previous reviewer and addressing them will really improve your application._

- [ ] _Check the comments under the review._

Cheers and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Please, remember to tag me in your question so I can receive the notification.

**_Please, do not open a new Pull Request for re-reviews. You should use the same Pull Request submitted for the first review, either valid or invalid unless it is requested otherwise._**

------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561)  you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

**Review by Cathella — APPROVED** [review]
## 🔥 Status: Approved 🟢

### Hi @Thermiee 👋🏻, 

Good job on making the requested changes 👍🏻.
Your project is complete!  There is nothing else to say other than... it's time to merge it :shipit: 
Congratulations! 🎉 


Cheers and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.


------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561)  you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

**Pull request merged**